### PR TITLE
Task 40 tests and scripts were added

### PR DESCRIPTION
**Pull request by igorchupin**
Task 40 tests and scripts were added

### CODE REVIEW

File: src/test/java/HalfDownloadTest.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    void halfDownloadTest() {
+        driver.get("https://www.seleniumeasy.com/test/bootstrap-download-progress-demo.html");
+        WebElement downloadButton = driver.findElement(By.id("cricle-btn"));
+        downloadButton.click();
+        do {
+            WebElement loadingPercent = driver.findElement(By.xpath("//div[@class='percenttext']"));
+            stringPercent = loadingPercent.getText();
+            percent = Integer.valueOf(stringPercent.substring(0, stringPercent.length()-1));
+        }
+        while (percent <=50);
```
Comment (EugenBorisik) [inline src/test/java/HalfDownloadTest.java]: Please create your custom ExpectedCondition

File: src/test/java/MultiselectTest.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.chrome.ChromeDriver;
+import org.openqa.selenium.support.ui.Select;
+
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import static org.junit.jupiter.api.Assertions.*;
+
+public class MultiselectTest {
```
Comment (EugenBorisik) [inline src/test/java/MultiselectTest.java:16]: Format code in this class

File: src/test/java/MultiselectTest.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        Select selectObject = new Select(multiSelectForm);
+        selectObject.selectByIndex(2);
+        selectObject.selectByValue("Texas");
+        selectObject.selectByVisibleText("Washington");
+        List<WebElement> selectedOptions = selectObject.getAllSelectedOptions();
+        for (int i = 0; i < selectedOptions.size(); i++) {
+            result = selectedOptions.get(i).getText().contains("New Jersey") || selectedOptions.get(i).getText().contains("Texas")
+                                                                        || selectedOptions.get(i).getText().contains("Washington");
+            if (result == false) {break;}
+        }
```
Comment (EugenBorisik) [inline src/test/java/MultiselectTest.java]: Can you rewrite this code with usage of java stream api?

File: src/test/java/TestsForAlerts.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    @DisplayName("Java_Script_Confirm_Box_Text_And_Approve")
+    void JavaScriptConfirmBoxTextApproveTest() {
+        WebElement buttonConfirm = driver.findElement(By.xpath("//button[@onclick='myConfirmFunction()']"));
+        buttonConfirm.click();
+        Alert alert = wait.until(ExpectedConditions.alertIsPresent());
+        String alertText = alert.getText();
+        alert.accept();
+
+        assertAll("Results",
+        ()-> assertEquals("You pressed OK!", driver.findElement(By.id("confirm-demo")).getText(),
```
Comment (EugenBorisik) [inline src/test/java/TestsForAlerts.java]: Format code

File: src/test/java/ListOfObjectTest.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                }
+            }
+            WebElement buttonNext = driver.findElement(By.id("example_next"));
+            buttonNext.click();
+        }
+        System.out.println("____________________________________________________________________________________");
+        System.out.println("Was found: " + peopleList.size() + " people;");
+        for (int z = 0; z < peopleList.size(); z++) {
+            peopleList.get(z).toString();
+        }
```
Comment (EugenBorisik) [inline src/test/java/ListOfObjectTest.java]: Move this code in separate method, which has parameters and returning value

File: src/test/java/ListOfObjectTest.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+    @Test
+    @Tag("List_Of_Objects")
+    @DisplayName("List of objects")
+    void listOfObjectsTest() {
+        ArrayList<People> peopleList = new ArrayList<People>();
+        driver.get("https://www.seleniumeasy.com/test/table-sort-search-demo.html");
+        WebElement selectEntries = driver.findElement(By.xpath("//select[@name='example_length']"));
+        Select selectObject = new Select(selectEntries);
+        selectObject.selectByValue("25");
```
Comment (EugenBorisik) [inline src/test/java/ListOfObjectTest.java]: 10

File: src/test/java/ListOfObjectTest.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        WebElement nextButton;
```
Comment (EugenBorisik) [inline src/test/java/ListOfObjectTest.java]: Remove

File: src/test/java/ListOfObjectTest.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        List<WebElement> buttonsCount = driver.findElements(By.xpath("//a[@class='paginate_button '] | " +
+                "//a[@class='paginate_button current']"));
+
+        for (int i = 1; i <= buttonsCount.size(); i++) {
+            List<WebElement> peopleListWeb = driver.findElements(By.xpath("//tbody/tr[@role='row']"));
+
+            for (int j = 0; j < peopleListWeb.size(); j++) {
+                String humanName = new String(peopleListWeb.get(j).findElement(By.xpath(String.format("//tr[%s]//td[1]", j+1))).getText());
```
Comment (EugenBorisik) [inline src/test/java/ListOfObjectTest.java]: "new String" is not needed

File: src/test/java/ListOfObjectTest.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+                String humanPosition = new String(peopleListWeb.get(j).findElement(By.xpath(String.format("//tr[%s]//td[2]", j+1))).getText());
```
Comment (EugenBorisik) [inline src/test/java/ListOfObjectTest.java]: "new String" is not needed

File: src/test/java/ListOfObjectTest.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+                String humanoffice = new String(peopleListWeb.get(j).findElement(By.xpath(String.format("//tr[%s]//td[3]", j+1))).getText());;
```
Comment (EugenBorisik) [inline src/test/java/ListOfObjectTest.java]: "new String" is not needed

File: src/test/java/ListOfObjectTest.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+                Integer age = new Integer(Integer.valueOf(peopleListWeb.get(j).findElement(By.xpath(String.format("//tr[%s]//td[4]", j+1))).getText()));
```
Comment (EugenBorisik) [inline src/test/java/ListOfObjectTest.java]: "new Integer" is not needed

File: src/test/java/ListOfObjectTest.java [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    @Tag("List_Of_Objects")
+    @DisplayName("List of objects")
+    void listOfObjectsTest() {
+        ArrayList<People> peopleList = new ArrayList<People>();
+        driver.get("https://www.seleniumeasy.com/test/table-sort-search-demo.html");
+        WebElement selectEntries = driver.findElement(By.xpath("//select[@name='example_length']"));
+        Select selectObject = new Select(selectEntries);
+        selectObject.selectByValue("25");
+        WebElement nextButton;
+        List<WebElement> buttonsCount = driver.findElements(By.xpath("//a[@class='paginate_button '] | " +
```
Comment (EugenBorisik) [inline src/test/java/ListOfObjectTest.java]: Bad locator

File: src/test/java/ListOfObjectTest.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+                "//a[@class='paginate_button current']"));
+
+        for (int i = 1; i <= buttonsCount.size(); i++) {
```
Comment (EugenBorisik) [inline src/test/java/ListOfObjectTest.java]: Replace with foreach

File: src/test/java/ListOfObjectTest.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+            List<WebElement> peopleListWeb = driver.findElements(By.xpath("//tbody/tr[@role='row']"));
+
+            for (int j = 0; j < peopleListWeb.size(); j++) {
```
Comment (EugenBorisik) [inline src/test/java/ListOfObjectTest.java]: Replace with foreach

File: src/test/java/ListOfObjectTest.java (hunk 8 above)
Comment (EugenBorisik) [inline src/test/java/ListOfObjectTest.java]: Now it looks like "bad hack", correct code will be:
.findElement(By.xpath(".//td[1]"))

**Review by EugenBorisik — CHANGES_REQUESTED** [review]

File: src/test/java/HalfDownloadTest.java (hunk 1 above)
Comment (igorchupin) [reply src/test/java/HalfDownloadTest.java]: Custom ExpectedCondition was created.
commit 48035613f843f33abbc671f222ddb3b35a6baa5d

File: src/test/java/MultiselectTest.java (hunk 3 above)
Comment (igorchupin) [reply src/test/java/MultiselectTest.java]: Stream API was used
commit 76ea1fd590496a135d0a872a55b8c697dc71c2f2

File: src/test/java/MultiselectTest.java (hunk 2 above)
Comment (igorchupin) [reply src/test/java/MultiselectTest.java:16]: I did it
commit 472c4dd27c36b2dc95ec03f816625444281a8a37

File: src/test/java/TestsForAlerts.java (hunk 4 above)
Comment (igorchupin) [reply src/test/java/TestsForAlerts.java]: Done.
commit 472c4dd27c36b2dc95ec03f816625444281a8a37

File: src/test/java/ListOfObjectTest.java (hunk 5 above)
Comment (igorchupin) [reply src/test/java/ListOfObjectTest.java]: Now, this code is a separate method.
commit be71aed623305902303353b0aea54704256fb6b1

File: src/test/java/ListOfObjectTest.java (hunk 8 above)
Comment (igorchupin) [reply src/test/java/ListOfObjectTest.java]: Removed.
commit be71aed623305902303353b0aea54704256fb6b1

File: src/test/java/ListOfObjectTest.java (hunk 9 above)
Comment (igorchupin) [reply src/test/java/ListOfObjectTest.java]: Removed.
commit be71aed623305902303353b0aea54704256fb6b1

File: src/test/java/ListOfObjectTest.java (hunk 10 above)
Comment (igorchupin) [reply src/test/java/ListOfObjectTest.java]: Removed.
commit be71aed623305902303353b0aea54704256fb6b1

File: src/test/java/ListOfObjectTest.java (hunk 11 above)
Comment (igorchupin) [reply src/test/java/ListOfObjectTest.java]: Removed.
commit be71aed623305902303353b0aea54704256fb6b1

File: src/test/java/ListOfObjectTest.java (hunk 12 above)
Comment (igorchupin) [reply src/test/java/ListOfObjectTest.java]: Locator was changed.
commit be71aed623305902303353b0aea54704256fb6b1

File: src/test/java/ListOfObjectTest.java (hunk 13 above)
Comment (igorchupin) [reply src/test/java/ListOfObjectTest.java]: Replaced with foreach.
commit be71aed623305902303353b0aea54704256fb6b1

File: src/test/java/ListOfObjectTest.java (hunk 14 above)
Comment (igorchupin) [reply src/test/java/ListOfObjectTest.java]: Replaced with foreach.
commit be71aed623305902303353b0aea54704256fb6b1

File: src/test/java/ListOfObjectTest.java (hunk 8 above)
Comment (igorchupin) [reply src/test/java/ListOfObjectTest.java]: This was changed. Thank you!
commit be71aed623305902303353b0aea54704256fb6b1

File: src/test/java/ListOfObjectTest.java (hunk 7 above)
Comment (igorchupin) [reply src/test/java/ListOfObjectTest.java]: This was removed.
commit be71aed623305902303353b0aea54704256fb6b1

File: src/test/java/ListOfObjectTest.java (hunk 6 above)
Comment (igorchupin) [reply src/test/java/ListOfObjectTest.java]: Value was changed.
commit be71aed623305902303353b0aea54704256fb6b1

File: src/test/java/HalfDownloadTest.java [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    void halfDownloadTest() {
+        driver.get("https://www.seleniumeasy.com/test/bootstrap-download-progress-demo.html");
+        WebElement downloadButton = driver.findElement(By.id("cricle-btn"));
+        downloadButton.click();
+        wait.until(new ExpectedCondition<Boolean>() {
+            public Boolean apply(WebDriver driver) {
+                WebElement loadingPercent = driver.findElement(By.xpath("//div[@class='percenttext']"));
+                stringPercent = loadingPercent.getText();
+                percent = Integer.valueOf(stringPercent.substring(0, stringPercent.length() - 1));
+                if (percent >= 50) return true;
```
Comment (EugenBorisik) [inline src/test/java/HalfDownloadTest.java]: Simplify: return percent >= 50;

File: src/test/java/MultiselectTest.java [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    void multiselectTest() {
+        boolean result = false;
+        driver.get("https://www.seleniumeasy.com/test/basic-select-dropdown-demo.html");
+        WebElement multiSelectForm = driver.findElement(By.id("multi-select"));
+        Select selectObject = new Select(multiSelectForm);
+        selectObject.selectByIndex(2);
+        selectObject.selectByValue("Texas");
+        selectObject.selectByVisibleText("Washington");
+        List<WebElement> selectedOptions = selectObject.getAllSelectedOptions();
+        Stream<WebElement> expected = selectedOptions.stream();
```
Comment (EugenBorisik) [inline src/test/java/MultiselectTest.java]: Remove this line

File: src/test/java/MultiselectTest.java [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+        result = expected.allMatch(x -> x.getText().contains("Texas") || x.getText().contains("Washington")
```
Comment (EugenBorisik) [inline src/test/java/MultiselectTest.java]: selectedOptions.stream().allMatch(...

Comment (EugenBorisik) [inline src/test/java/MultiselectTest.java]: Create collection of options (named expectedOptions f.e.) in any place before stream.
Then simplify:
selectedOptions.stream().map(WebElement::getText).allMatch(expectedOptions::contains);
And place this code just in assertTrue method

File: src/test/java/MultiselectTest.java [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+
+    @AfterEach
+    public void afterTest() {
+        driver.quit();
+    }
+
+    @Test
+    @Tag("Multiselect")
+    void multiselectTest() {
+        boolean result = false;
```
Comment (EugenBorisik) [inline src/test/java/MultiselectTest.java]: Remove

File: src/test/java/ListOfObjectTest.java [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+                    People tempPeople = new People(humanName, humanPosition, humanOffice, age, salaryInt);
+                    peopleList.add(tempPeople);
+                }
+            }
+            WebElement buttonNext = driver.findElement(By.id("example_next"));
+            buttonNext.click();
+        }
+        return peopleList;
+    }
+}
```
Comment (EugenBorisik) [inline src/test/java/ListOfObjectTest.java:69]: Just want to provide you example for this task using only stream api:
```
public List<People> getEmployees(int age, int salary) {
    WebElement selectEntries = driver.findElement(By.xpath("//select[@name='example_length']"));
    Select selectObject = new Select(selectEntries);
    selectObject.selectByValue("10");

    String paginationButton = "//a[contains(@class, 'paginate_button') and text()='%d']";
    List<WebElement> paginationButtons = driver.findElements(By.cssSelector("span>a.paginate_button"));
    return IntStream.rangeClosed(1, paginationButtons.size())
        .mapToObj(i -> driver.findElement(By.xpath(String.format(paginationButton, i))))
        .map(toEmployeesOnCurrentPage(age, salary))
        .flatMap(Collection::stream)
        .collect(Collectors.toList());
  }

  private Function<WebElement, List<People>> toEmployeesOnCurrentPage(int age, int salary) {
    return button -> {
      button.click();

      List<WebElement> rows = driver.findElements(By.xpath("//table[@id='example']//tbody/tr"));
      return rows.stream()
          .map(row -> {
            String currentName = row.findElement(By.xpath(".//td[1]")).getText();
            String currentPosition = row.findElement(By.xpath(".//td[2]")).getText();
            String currentOffice = row.findElement(By.xpath(".//td[3]")).getText();
            Integer currentAge = Integer.valueOf(row.findElement(By.xpath(String.format(".//td[4]"))).getText());
            String currentSalary = row.findElement(By.xpath(".//td[6]")).getText();
            currentSalary = currentSalary.substring(1, currentSalary.length() - 2).replace(",", "");
            Integer salaryInt = Integer.valueOf(currentSalary);

            return new People(currentName, currentPosition, currentOffice, currentAge, salaryInt);
          })
          .filter(employee -> employee.getAge() > age && employee.getSalary() <= salary)
          .collect(Collectors.toList());
    };
  }
```

**Review by EugenBorisik — CHANGES_REQUESTED** [review]

File: src/test/java/HalfDownloadTest.java (hunk 15 above)
Comment (igorchupin) [reply src/test/java/HalfDownloadTest.java]: This was changed.
commit 582e14338eba808c7b8b3e9cec8e8304f329e004

File: src/test/java/ListOfObjectTest.java (hunk 19 above)
Comment (igorchupin) [reply src/test/java/ListOfObjectTest.java:69]: Thank you I need to read more about streams =(((

File: src/test/java/MultiselectTest.java (hunk 18 above)
Comment (igorchupin) [reply src/test/java/MultiselectTest.java]: Removed
commit 582e14338eba808c7b8b3e9cec8e8304f329e004

File: src/test/java/MultiselectTest.java (hunk 16 above)
Comment (igorchupin) [reply src/test/java/MultiselectTest.java]: Removed
commit 582e14338eba808c7b8b3e9cec8e8304f329e004

File: src/test/java/MultiselectTest.java (hunk 17 above)
Comment (igorchupin) [reply src/test/java/MultiselectTest.java]: Changed
commit 582e14338eba808c7b8b3e9cec8e8304f329e004

Comment (igorchupin) [reply src/test/java/MultiselectTest.java]: Changed
commit 582e14338eba808c7b8b3e9cec8e8304f329e004

**Review by EugenBorisik — APPROVED** [review]

**Pull request merged**